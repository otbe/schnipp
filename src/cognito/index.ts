import { CognitoUserPoolTriggerEvent } from 'aws-lambda';
import { IHandler } from '../';

export type CognitoUserPoolHandler = IHandler<
  CognitoUserPoolTriggerEvent,
  CognitoUserPoolTriggerEvent
>;
