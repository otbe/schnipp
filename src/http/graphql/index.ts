export { GraphQLController } from './GraphQLController';
export { Query } from './decorators/query';
export { Mutation } from './decorators/mutation';
export { ResolveField } from './decorators/resolveField';
export { ResolveType } from './decorators/resolveType';
export { Header } from './decorators/header';
export { Meta } from './decorators/meta';
export { Use } from './decorators/use';
