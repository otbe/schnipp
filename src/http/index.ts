import {
  APIGatewayEvent,
  APIGatewayProxyResult,
  CustomAuthorizerEvent,
  CustomAuthorizerResult
} from 'aws-lambda';
import { IHandler } from '..';

export type APIGatewayHandler = IHandler<
  APIGatewayEvent,
  APIGatewayProxyResult
>;

export type AuthorizeHandler = IHandler<
  CustomAuthorizerEvent,
  CustomAuthorizerResult
>;

export { AuthPolicy, HttpVerb, Effect } from './utils/AuthPolicy';
export { createResponse } from './utils/createResponse';
export { MetaData } from './utils/MetaData';
export {
  DecoratedExceptionFilter,
  ExceptionFilter
} from './utils/ExceptionFilter';
export { Guard } from './utils/Guard';
export { Catch } from './decorators/catch';
export { RestController } from './RestController';

export {
  InjectedParam,
  MethodMetaData,
  ParamInjector
} from './factories/createHttpMethodDecorator';

export {
  GET,
  POST,
  DELETE,
  PUT,
  PATCH,
  OPTIONS,
  HEAD
} from './decorators/methods';

export { Path } from './decorators/path';
export { Param } from './decorators/param';
export { Query } from './decorators/query';
export { StatusCode } from './decorators/statusCode';
export { Context } from './decorators/context';
export { Body } from './decorators/body';
export { Header } from './decorators/header';
export { Meta } from './decorators/meta';
export { Use } from './decorators/use';

export { HttpException } from './exceptions/HttpException';
export { NotFoundException } from './exceptions/NotFoundException';
export { ForbiddenException } from './exceptions/ForbiddenException';
