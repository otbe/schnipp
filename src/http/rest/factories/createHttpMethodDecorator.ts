import * as urlJoin from 'url-join';
import { APIGatewayEvent } from 'aws-lambda';
import { Newable } from 'simple-ts-di';
import { HttpVerb } from '../../utils/AuthPolicy';
import { DecoratedExceptionFilter } from '../../utils/ExceptionFilter';
import { Guard } from '../../utils/Guard';
import { MetaData } from '../../utils/MetaData';
import { DefaultExecutionContext } from '../../utils/ExecutionContext';

const METHODS = Symbol('methods');

export type ParamInjector<T> = {
  <Context = DefaultExecutionContext>(
    e: APIGatewayEvent,
    metaData: MetaData,
    context: Context
  ): T;
};

export type InjectedParam<T> = {
  index: number;
  injector: ParamInjector<T>;
};

export type MethodMetaData = {
  statusCode?: number;
  method?: HttpVerb;
  path: string;
  propertyKey: string;
  headers: { [key: string]: string };
  paramaters: Array<InjectedParam<any>>;
  filters: Array<DecoratedExceptionFilter>;
  guards: Array<Newable<Guard>>;
  metaData: MetaData;
};

export const createHttpMethodDecorator = (method: HttpVerb) => (
  path: string = '/'
): MethodDecorator => <T>(
  target: any,
  propertyKey: string | symbol,
  descriptor: TypedPropertyDescriptor<T>
) => {
  const methodMetaData = getMethodMetaData(propertyKey.toString(), target);
  if (methodMetaData.statusCode == null) {
    methodMetaData.statusCode = method === 'POST' ? 201 : 200;
  }
  methodMetaData.method = method;
  methodMetaData.path = path;
  return descriptor;
};

export const getMethodMetaData = (propertyKey: string, controller: any) => {
  let methods: Array<MethodMetaData> | null = Reflect.getMetadata(
    METHODS,
    controller
  );

  if (methods == null) {
    methods = [];
    Reflect.defineMetadata(METHODS, methods, controller);
  }

  let data = methods.find(x => x.propertyKey === propertyKey);

  if (data == null) {
    data = {
      path: '/',
      propertyKey,
      filters: [],
      paramaters: [],
      guards: [],
      headers: {},
      metaData: {}
    };
    methods.push(data);
  }

  return data;
};

export const findMethod = (
  basePath: string,
  resourcePath: string,
  method: string,
  controller: any
) => {
  const m: Array<MethodMetaData> = Reflect.getMetadata(METHODS, controller);

  const res = m.find(
    x =>
      x.method === method &&
      urlJoin(basePath, x.path!.replace(/\/$/, '')) === resourcePath
  );

  if (res == null) {
    return m.find(x => x.method === '*');
  }

  return res;
};
