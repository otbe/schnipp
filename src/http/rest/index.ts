export { RestController } from './RestController';

export {
  InjectedParam,
  MethodMetaData,
  ParamInjector
} from './factories/createHttpMethodDecorator';

export {
  GET,
  POST,
  DELETE,
  PUT,
  PATCH,
  OPTIONS,
  HEAD
} from './decorators/methods';

export { Path } from './decorators/path';
export { Param } from './decorators/param';
export { Query } from './decorators/query';
export { StatusCode } from './decorators/statusCode';
export { Context } from './decorators/context';
export { Body } from './decorators/body';
export { Header } from './decorators/header';
export { Meta } from './decorators/meta';
export { Use } from './decorators/use';

export { HttpException } from './exceptions/HttpException';
export { NotFoundException } from './exceptions/NotFoundException';
export { ForbiddenException } from './exceptions/ForbiddenException';
