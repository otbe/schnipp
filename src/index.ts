import { Context, Handler } from 'aws-lambda';
import { ContainedType, getFromContainer } from './container';

export type IHandler<E, R = void> = {
  handle(event: E, context: Context): Promise<R> | R;
};

let setupPromise: Promise<any> | null = null;
export const createHandler = <E, R>(
  handler: ContainedType<IHandler<E, R>>,
  setupFn?: () => Promise<any>
): Handler<E, R> => async (event: E, context: Context) => {
  context.callbackWaitsForEmptyEventLoop = false;

  if (setupFn) {
    if (setupPromise == null) {
      setupPromise = setupFn();
    }

    await setupPromise;
  }

  const handlerInstance = getFromContainer(handler);
  return await handlerInstance.handle(event, context);
};

export { useContainer } from './container';
