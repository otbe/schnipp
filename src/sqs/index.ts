import { SQSEvent } from 'aws-lambda';
import { IHandler } from '../';

export type SQSHandler = IHandler<SQSEvent, void>;
