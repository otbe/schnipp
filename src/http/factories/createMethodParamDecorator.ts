import { ParamInjector, getMethodMetaData } from './createHttpMethodDecorator';

export const createMethodParameterDecorator = <T>(
  injector: ParamInjector<T>
): ParameterDecorator => (
  target: Object,
  propertyKey: string | symbol,
  index: number
) => {
  const methodMetaData = getMethodMetaData(propertyKey.toString(), target);

  methodMetaData.paramaters.push({
    index,
    injector
  });
};
