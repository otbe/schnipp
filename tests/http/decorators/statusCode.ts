import 'reflect-metadata';
import { StatusCode, GET, POST } from '../../../src/http';
import { getMethodMetaData } from '../../../src/http/factories/createHttpMethodDecorator';

describe('@StatusCode', () => {
  it('should add statusCode ', () => {
    class Test {
      @StatusCode(245)
      test() {}
    }

    const f = new Test();

    expect(getMethodMetaData('test', f)).toMatchSnapshot();
  });

  it('should coop with methods ', () => {
    class Test {
      @GET('/')
      @StatusCode(245)
      test() {}

      @POST('/bar')
      @StatusCode(202)
      test2() {}
    }

    const f = new Test();
    expect(getMethodMetaData('test', f)).toMatchSnapshot();
    expect(getMethodMetaData('test2', f)).toMatchSnapshot();
  });
});
