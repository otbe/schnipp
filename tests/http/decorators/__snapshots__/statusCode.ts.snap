// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@StatusCode should add statusCode  1`] = `
Object {
  "filters": Array [],
  "guards": Array [],
  "headers": Object {},
  "metaData": Object {},
  "paramaters": Array [],
  "path": "/",
  "propertyKey": "test",
  "statusCode": 245,
}
`;

exports[`@StatusCode should coop with methods  1`] = `
Object {
  "filters": Array [],
  "guards": Array [],
  "headers": Object {},
  "metaData": Object {},
  "method": "GET",
  "paramaters": Array [],
  "path": "/",
  "propertyKey": "test",
  "statusCode": 245,
}
`;

exports[`@StatusCode should coop with methods  2`] = `
Object {
  "filters": Array [],
  "guards": Array [],
  "headers": Object {},
  "metaData": Object {},
  "method": "POST",
  "paramaters": Array [],
  "path": "/bar",
  "propertyKey": "test2",
  "statusCode": 202,
}
`;
